### ISSUE TITLE
Generate ESM imports for browser targets

### ISSUE THREAD

**Issue by nex3** [google/dart_cli_pkg#117]
If the user wants to distribute a browser-compatible library, the CommonJS `require()`s we generate aren't ideal. They can work with the aid of a bundler, but they can't be loaded directly by the browser. If the user has chosen to generate a browser-compatible library, we should generate a `<pkg>.browser.js` that uses ESM for all its loads. This also means that we should inject _all_ loads in this case rather than including any directly in the `.dart.js` preamble.

Specifically:

* We should add a `pkg.jsSupportsBrowser` option that indicates whether the `jsModuleMainLibrary` can be loaded on a browser. It should default to true iff `pkg.jsModuleMainLibrary` is set and `pkg.jsRequires` contains a require whose target is `JsRequireTarget.browser` or `JsRequireTarget.all`. JS generation should throw an error if `pkg.jsSupportsBrowser` is set to true and `pkg.jsModuleMainLibrary` is unset.

* If `pkg.jsSupportsBrowser` is true and any dependencies are detected (either through `pkg.jsRequires` or extracted from the compiled code), all dependencies should be injected and a `<pkg>.browser.js` file should be generated which loads the dependencies and `<pkg>.dart.js` using ESM imports.

**Comment by nex3** [comment]
On consideration, I think it's not worth having a separate `pkg.jsSupportsBrowser` config—we should just generate a browser target based on whether it would be set by default. If users don't want that, they can configure that using `pkg.jsRequires`.

**Comment by nex3** [comment]
@jerivas I've run into a roadblock here. It's pretty easy to update the dependency imports to use ESM, but I'm not sure if there's a good way to update `sass.dart.js` itself to expose its members as ESM _while also being loadable as CommonJS_. The JS file is very large so I don't want to fully duplicate it, but I also can't guard ESM exports with conditionals. Do you know of anything I can generate here that could make this wok?

**Comment by nex3** [comment]
All right, I think I figured it out. We'll need to explicitly enumerate all exports for the browser, but that's not too steep a price to pay.

**Comment by nex3** [comment]
So here's an interesting quandary: in order to make this work, I need to use a top-level `await import("./sass.dart.js")`. This is fine when running JS raw in the browser, since ESM allows top-level await, but it's iffier with bundlers—Webpack, pertinently, still locks top-level await behind an experiment. On the other hand, once a user is running a bundler CommonJS imports are totally fine for them since they're resolved at build time anyway.

It's easy enough to write two different entrypoints for the browser, one that uses CommonJS and one that uses ESM. But the trick is figuring out how to expose these to users. There's only one [`"browser"` condition](https://nodejs.org/api/packages.html#community-conditions-definitions) in the export field, so there isn't a way to say "use this file for CommonJS on the browser and this one for ESM on the browser". But those fields exist to be consumed by tools anyway, so maybe the correct answer is to always point to the CommonJS file and just document "there's also this ESM file if you want to load it directly in the browser".

@jerivas WDYT?

**Comment by jerivas** [comment]
That sounds reasonable to me. To be honest I have limited experience with npm packaging, so I'd like to ask @jgerigmeyer for help. In general, I'm ok deferring to anyone with stronger opinions on this particular issue

**Comment by nex3** [comment]
I'm also not very familiar with JS packaging, especially when it comes to serving code to the browser, so I would definitely appreciate a more experienced perspective.

**Comment by jerivas** [comment]
Another option that came to mind is replacing CommonJS with ESM now that node supports it natively, but some basic googling seems to indicate it's not straight forward

**Comment by jgerigmeyer** [comment]
In this case I might be tempted to skip the `browser` field entirely, and rely on `main` (CJS) and `module` (ESM) (and maybe also `exports.import` and `exports.require`). As a consumer, I think I'd expect anything in `main` or `module` to expose an API without side-effects (which is I think the primary use-case?), where something imported from `browser` would have side-effects, likely setting a `window.sass` var or something. So if there's a `browser` field, I think that should maybe be a separate IIFE bundle -- which might not play as nicely with top-level `await`.

If you use `main` and `module`, it's true that Webpack might default to the ESM `module` with top-level await, but users should be able to work around that either by enabling `topLevelAwait` or by directly importing the CJS file. So I think that'd be my leaning...

I'm also fine with your approach -- which I think is just leaving the CJS bundle (`main`) as the default, but also exporting an ESM bundle that users can import directly? That's essentially the purpose of the `module` field, but I understand why it might make sense not to use it for now.

tl;dr -- JS module packaging is still a headache

**Comment by nex3** [comment]
I don't think we can reliably use `main` vs `module` here (or equivalently `import` vs `require` in `package.exports`) because we specifically need to have separate entrypoints for the browser and Node. Node needs to load the `fs` and `utils` modules, which don't exist on the browser, and although we don't currently need to load any browser-only modules the infrastructure needs to remain open to that possibility.

So to reiterate, the plan is:

* Set `package.exports.default` to the Node CJS entrypoint (`sass.node.js`).
* Set `package.exports.browser` to the Browser CJS entrypoint (`sass.browser.cjs`).
* Add a `sass.browser.js` Browser ESM entrypoint that's not explicitly listed in `package.json` for users to `import` directly.

**Comment by nex3** [comment]
A further complication has arisen. There's a middle ground between Webpack's fully-compiled toolchain and a raw `import` in a browser: [JSPM](https://jspm.org/). Their generator tool creates import maps that make it possible to load modules _by name_ in the browser, but it doesn't do any further bundling and so is unable to unwrap a CJS entrypoint.

The good news is that I think I have a way to load the package via ESM _without_ top-level `await`. That should allow Webpack to consume it happily enough while still allowing it to work without bundling in the browser.

So the new plan is:

* Set `package.exports.default` to the Node CJS entrypoint (`sass.node.js`).
* Set `package.exports.browser` to the Browser ESM entrypoint (`sass.browser.js`).
* Do not provide a Browser CJS entrypoint at all.

**Comment by jgerigmeyer** [comment]
> I don't think we can reliably use `main` vs `module` here (or equivalently `import` vs `require` in `package.exports`) because we specifically need to have separate entrypoints for the browser and Node. Node needs to load the `fs` and `utils` modules, which don't exist on the browser, and although we don't currently need to load any browser-only modules the infrastructure needs to remain open to that possibility.

@nex3 Ah yes, that makes sense! Thanks for clarifying -- I was missing that crucial point.

> So the new plan is:
> 
> * Set `package.exports.default` to the Node CJS entrypoint (`sass.node.js`).
> * Set `package.exports.browser` to the Browser ESM entrypoint (`sass.browser.js`).
> * Do not provide a Browser CJS entrypoint at all.

That sounds ideal -- any modern bundler should be able to handle ESM for the browser, especially without top-level `await`. 🚀

**Comment by nex3** [comment]
Apparently I completely overlooked the fact that [nested conditions](https://nodejs.org/api/packages.html#nested-conditions) exist in `package.exports`, so we don't need to be _nearly_ as clever as I thought. We can just generate four separate  entrypoints for CJS/ESM cross Node/browser.

**Comment by nex3** [comment]
Fixed by https://github.com/google/dart_cli_pkg/pull/118
